src/audio/dai-zephyr.c

```diff
@@ -1461,6 +1461,37 @@ int dai_zephyr_multi_endpoint_copy(struct dai_data **dd, struct comp_dev *dev,
 	return 0;
 }
 
+static void set_new_local_buffer(struct dai_data *dd, struct comp_dev *dev)
+{
+	struct comp_buffer __sparse_cache *dma_buf = buffer_acquire(dd->dma_buffer);
+	struct comp_buffer __sparse_cache *local_buf;
+	uint32_t dma_fmt = audio_stream_get_frm_fmt(&dma_buf->stream);
+	uint32_t local_fmt;
+
+	buffer_release(dma_buf);
+
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		dd->local_buffer = list_first_item(&dev->bsource_list,
+						   struct comp_buffer,
+						   sink_list);
+	else
+		dd->local_buffer = list_first_item(&dev->bsink_list,
+						   struct comp_buffer,
+						   source_list);
+
+	local_buf = buffer_acquire(dd->local_buffer);
+	local_fmt = audio_stream_get_frm_fmt(&local_buf->stream);
+	buffer_release(local_buf);
+
+	dd->process = pcm_get_conversion_function(local_fmt, dma_fmt);
+
+	if (!dd->process) {
+		comp_err(dev, "converter function NULL: local fmt %d dma fmt %d\n",
+			 local_fmt, dma_fmt);
+		dd->local_buffer = NULL;
```

Comment:
can we use a temporary variable in lines 1473-1480 above and only assign `dd->local_buffer` after this check?

Reply:
@lyakh Took a look at this, but actually I'd prefer not to. We know at this point already that the old value of "dd->local_buffer" is invalid and will be sooner freed from the heap, so we want to erase in any case. I.e. even if no converter is found, we want to reset dd->local_buffer.

Reply:
sorry, maybe I didn't explain well. This function is only called `if (!dd->local_buffer)` right? Then it is assigned a nun-NULL value either in line 1474 or in line 1478 above. But then we check if we can find a conversion function and if not - set that pointer to NULL *again*. Whereas we could simply avoid setting it to something we cannot use anyway completely. That way at no point of time would it be set to something useless. I.e.
```
	struct comp_buffer *buf;
	...
	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
		buf = list_first_item(&dev->bsource_list,
						   struct comp_buffer,
						   sink_list);
	else
		buf = list_first_item(&dev->bsink_list,
						   struct comp_buffer,
						   source_list);

	local_buf = buffer_acquire(buf);
	local_fmt = audio_stream_get_frm_fmt(&local_buf->stream);
	buffer_release(local_buf);

	dd->process = pcm_get_conversion_function(local_fmt, dma_fmt);

	if (!dd->process)
		comp_err(dev, "converter function NULL: local fmt %d dma fmt %d\n",
			 local_fmt, dma_fmt);

	dd->local_buffer = buf;
```

Reply:
Right, now I get your point @lyakh . Let me do a small fixup to this in separate PR given I got a green run on the CI and ok from reviewers and we can't of really need to have this fix asap in the tree.